File: scripts/battery_analysis.py
"""
Battery Analysis Script (Python)
@name Battery Analysis (Python)
@version 1.0.0
@author Toolbox Team
@description Advanced battery analysis with NumPy statistics, trend analysis, and health scoring
@source https://github.com/GameTec-live/ChameleonUltra
"""

import asyncio
import numpy as np
from js import ToolboxAPI

# Import constants from Chameleon Ultra device extension
# (Device extension exports these, not hardcoded here!)
CMD_GET_BATTERY_INFO = ToolboxAPI.ChameleonUltra.CMD_GET_BATTERY_INFO

async def get_battery():
    """Get battery info using device extension"""
    result = await device_cmd(CMD_GET_BATTERY_INFO, None)
    if result and result['data']:
        data = result['data']
        voltage = (data[1] << 8) | data[0]
        percentage = data[2] if len(data) > 2 else 0
        return {'voltage': voltage, 'percentage': percentage}
    return None

async def analyze_battery():
    """
    Comprehensive battery analysis
    """
    print("=" * 60)
    print("🔋 BATTERY ANALYSIS")
    print("=" * 60)

    # Check connection
    if not is_connected():
        print("❌ Not connected to device")
        return

    # Collect 15 battery samples
    print("\n📊 Collecting 15 samples over 7.5 seconds...")

    voltages = []
    percentages = []

    for i in range(15):
        try:
            battery = await get_battery()

            if battery:
                voltages.append(battery['voltage'])
                percentages.append(battery['percentage'])

                if (i + 1) % 5 == 0:
                    print(f"  Progress: {i + 1}/15 samples...")

            await asyncio.sleep(0.5)

        except Exception as e:
            print(f"❌ Error: {str(e)}")
            break

    if len(voltages) < 5:
        print("❌ Insufficient data collected")
        return

    # Convert to NumPy arrays
    v_arr = np.array(voltages)
    p_arr = np.array(percentages)

    print(f"\n✓ Collected {len(voltages)} samples")

    # === STATISTICS ===
    print("\n" + "=" * 60)
    print("📊 BATTERY STATISTICS")
    print("=" * 60)

    print(f"\nVoltage:")
    print(f"  Mean:     {np.mean(v_arr):.1f} mV")
    print(f"  Median:   {np.median(v_arr):.1f} mV")
    print(f"  Std Dev:  {np.std(v_arr):.1f} mV")
    print(f"  Range:    {np.min(v_arr)} - {np.max(v_arr)} mV")

    print(f"\nBattery Level:")
    print(f"  Mean:     {np.mean(p_arr):.1f}%")
    print(f"  Median:   {np.median(p_arr):.1f}%")
    print(f"  Current:  {p_arr[-1]}%")

    # === TREND ANALYSIS ===
    print("\n" + "=" * 60)
    print("📈 TREND ANALYSIS")
    print("=" * 60)

    # Calculate trend (linear regression)
    trend = np.polyfit(range(len(v_arr)), v_arr, 1)[0]

    if abs(trend) < 0.5:
        status = "🟢 STABLE"
        print(f"\n  Status: {status}")
        print(f"  Change: {trend:.3f} mV/sample (negligible)")
    elif trend > 0:
        status = "📈 CHARGING"
        print(f"\n  Status: {status}")
        print(f"  Rate: +{trend:.2f} mV/sample")
    else:
        status = "📉 DRAINING"
        print(f"\n  Status: {status}")
        print(f"  Rate: {trend:.2f} mV/sample")

    # === HEALTH ASSESSMENT ===
    print("\n" + "=" * 60)
    print("🏥 HEALTH ASSESSMENT")
    print("=" * 60)

    level = np.mean(p_arr)
    stability = max(0, 100 - np.std(v_arr))
    health_score = level * 0.7 + stability * 0.3

    print(f"\n  Battery Level:    {level:.1f}/100")
    print(f"  Stability:        {stability:.1f}/100")
    print(f"  Overall Score:    {health_score:.1f}/100")

    if health_score >= 80:
        verdict = "✅ EXCELLENT"
    elif health_score >= 60:
        verdict = "🟡 GOOD"
    elif health_score >= 40:
        verdict = "🟠 FAIR"
    else:
        verdict = "🔴 POOR"

    print(f"\n  Verdict: {verdict}")

    # === RECOMMENDATIONS ===
    print("\n" + "=" * 60)
    print("💡 RECOMMENDATIONS")
    print("=" * 60)

    if level < 20:
        print("  ⚡ Charge battery immediately")
    elif level < 50:
        print("  🔋 Consider charging soon")

    if np.std(v_arr) > 50:
        print("  📡 Check BLE connection stability")

    if trend < -1:
        print("  ⏰ Battery draining quickly")

    print("\n" + "=" * 60)
    print("✓ Analysis complete!")

# Run the analysis
await analyze_battery()

File: scripts/battery_analysis.lua
--[[
Battery Analysis Script (Lua)
@name Battery Analysis (Lua)
@version 1.0.0
@author Toolbox Team
@description Battery analysis with statistics, trend analysis, and health scoring in pure Lua
@source https://github.com/GameTec-live/ChameleonUltra
--]]

-- Import constants from Chameleon Ultra device extension
-- (Device extension exports these, not hardcoded here!)
local CMD_GET_BATTERY_INFO = js.global.ToolboxAPI.ChameleonUltra.CMD_GET_BATTERY_INFO

-- Helper function to get battery info using device extension
local function get_battery()
    local result = device_cmd(CMD_GET_BATTERY_INFO, nil)
    if result and result.data then
        local voltage = result.data[1] + (result.data[2] * 256)
        local percentage = result.data[3] or 0
        return {voltage = voltage, percentage = percentage}
    end
    return nil
end

-- Battery Analysis Script in Lua
print("============================================================")
print("🔋 BATTERY ANALYSIS (LUA)")
print("============================================================")

-- Check connection
if not is_connected() then
    print("❌ Not connected to device")
    return
end

-- Note: Sample collection must be done from JavaScript side due to async limitations
-- This script focuses on analysis of pre-collected data

-- For now, let's do a simple demo with a few samples
print("")
print("📊 Collecting battery samples...")
print("  (Using device bridge functions)")

local samples = {}
local sample_count = 5

-- Collect samples (these are synchronous Lua calls that wrap async JS)
for i = 1, sample_count do
    local battery = get_battery()

    if battery then
        table.insert(samples, {
            voltage = battery.voltage,
            percentage = battery.percentage
        })
        print(string.format("  Sample %d: %dmV (%d%%)", i, battery.voltage, battery.percentage))
    end
end

if #samples < 3 then
    print("❌ Insufficient data collected")
    return
end

print(string.format("\n✓ Collected %d samples", #samples))

-- Calculate statistics
local function mean(t, key)
    local sum = 0
    for _, v in ipairs(t) do
        sum = sum + v[key]
    end
    return sum / #t
end

local function min(t, key)
    local m = t[1][key]
    for _, v in ipairs(t) do
        if v[key] < m then m = v[key] end
    end
    return m
end

local function max(t, key)
    local m = t[1][key]
    for _, v in ipairs(t) do
        if v[key] > m then m = v[key] end
    end
    return m
end

local function stddev(t, key)
    local m = mean(t, key)
    local sum = 0
    for _, v in ipairs(t) do
        sum = sum + (v[key] - m)^2
    end
    return math.sqrt(sum / #t)
end

-- Voltage statistics
local v_mean = mean(samples, 'voltage')
local v_min = min(samples, 'voltage')
local v_max = max(samples, 'voltage')
local v_std = stddev(samples, 'voltage')

-- Percentage statistics
local p_mean = mean(samples, 'percentage')
local p_current = samples[#samples].percentage

print("")
print("============================================================")
print("📊 BATTERY STATISTICS")
print("============================================================")

print("")
print("Voltage:")
print(string.format("  Mean:     %.1f mV", v_mean))
print(string.format("  Min:      %d mV", v_min))
print(string.format("  Max:      %d mV", v_max))
print(string.format("  Std Dev:  %.1f mV", v_std))
print(string.format("  Range:    %d mV", v_max - v_min))

print("")
print("Battery Level:")
print(string.format("  Mean:     %.1f%%", p_mean))
print(string.format("  Current:  %d%%", p_current))

-- Trend analysis (simple linear regression)
local function linear_trend(data, key)
    local n = #data
    local sum_x = 0
    local sum_y = 0
    local sum_xy = 0
    local sum_x2 = 0

    for i, item in ipairs(data) do
        local x = i - 1
        local y = item[key]
        sum_x = sum_x + x
        sum_y = sum_y + y
        sum_xy = sum_xy + (x * y)
        sum_x2 = sum_x2 + (x * x)
    end

    local slope = (n * sum_xy - sum_x * sum_y) / (n * sum_x2 - sum_x * sum_x)
    return slope
end

local trend = linear_trend(samples, 'voltage')

print("")
print("============================================================")
print("📈 TREND ANALYSIS")
print("============================================================")

print("")
if math.abs(trend) < 0.5 then
    print("  Status: 🟢 STABLE")
    print(string.format("  Change: %.3f mV/sample (negligible)", trend))
elseif trend > 0 then
    print("  Status: 📈 CHARGING")
    print(string.format("  Rate: +%.2f mV/sample", trend))
else
    print("  Status: 📉 DRAINING")
    print(string.format("  Rate: %.2f mV/sample", trend))
end

-- Health assessment
print("")
print("============================================================")
print("🏥 HEALTH ASSESSMENT")
print("============================================================")

local level_score = math.min(100, p_mean)
local stability_score = math.max(0, 100 - v_std)
local health_score = level_score * 0.7 + stability_score * 0.3

print("")
print(string.format("  Battery Level:    %.1f/100", level_score))
print(string.format("  Stability:        %.1f/100", stability_score))
print(string.format("  Overall Score:    %.1f/100", health_score))

local verdict
if health_score >= 80 then
    verdict = "✅ EXCELLENT"
elseif health_score >= 60 then
    verdict = "🟡 GOOD"
elseif health_score >= 40 then
    verdict = "🟠 FAIR"
else
    verdict = "🔴 POOR"
end

print("")
print("  Verdict: " .. verdict)

-- Recommendations
print("")
print("============================================================")
print("💡 RECOMMENDATIONS")
print("============================================================")

local has_recommendations = false

if p_mean < 20 then
    print("  ⚡ Charge battery immediately")
    has_recommendations = true
elseif p_mean < 50 then
    print("  🔋 Consider charging soon")
    has_recommendations = true
end

if v_std > 50 then
    print("  📡 Check BLE connection stability")
    has_recommendations = true
end

if trend < -1 then
    print("  ⏰ Battery draining quickly")
    has_recommendations = true
end

if not has_recommendations then
    print("  ✅ Battery is healthy, no action needed")
end

print("")
print("============================================================")
print("✓ Analysis complete!")
print("============================================================")
